third_party/lua/luadebug.lua

```diff
@@ -464,12 +471,19 @@ local function cmd_add_breakpoint(bps)
     local file = #fullfile > 0 and fullfile or debuggee
     local normfile = normalize_path(file)
 
+    -- save direct hash (for faster lookup): line -> file
```

Comment:
I am not sure, but isn't that premature optimization?

Reply:
We've discussed it verbally with @tsafin, and the feature is actually an optimization for `has_breakpoint()`. But I'm still not sure that it works, as in many cases `line` is valid for almost all files, the file lookup will be executed for each file anyway.

Reply:
I still don't understand why do we leave a table with two different field types. I don't see that it is required for something besides questionable convenience (IMO, the need to sync the values eliminates it).

Reply:
Sigh. We have discussed it multiple times offline, but let me reiterate a few topics here:

* Let assume we need to handle pair values {`filename`, `linenumber`} we may equally save it as filename in 1st dimension, and then linenumber in 2nd dimension, or in reverse linenumber in 1st dimension, and filename in second;
* From usability point of view (filename, linenumber) order is preferred to show breakpoints saved in natural order (sorted by filenames first, then by line numbers) for the 
* From a performance and common sense point of view we may prefer to first check for a line number, as it's providing better selectivity, and less chance of conflicts requiring checking of a 2nd condition (given the assumption that we frequently debug a single script, with a few breakpoints activated, and a thousand other lines which would not require any stop during execution - i.e. majority, 99.999% of lines should interrupt in debugger).
* Guesstimate was that we would slow down execution 100x times comparing to the full speed JIT (more precise data will be shown later), and would rather to select fast track out if the line processed is not requiring any debugger attention.

Benchmarking approach
================

To estimate the biggest overhead debugger could introduce into cpu-bound algorithm, we have selected a classical Mandelbrot implementation by Mike Pall and wrapped it out inside a code which was activating debugger-like code (copied almost verbatim from `luadebug.lua`).

Here is the code https://gist.github.com/tsafin/da66ac0c992c6c4442d6e5572ae6882a#file-mandelbrot-dbg-2-lua

We see there are several cases collected:
- Mandelbrot with `jit.on()` and not debugger hook enabled. That's the fastest mode, and will be used as a basis for comparison with others;
- `jit.on()` but with debugger hook enabled. We have discovered the harder way that we had to flush all JIT traces before enabling `debug.sethook` otherwise some already translated traces will continue to execute in full speed, without calling the debugger. But once properly activated, one dramatically reduces generic speed of execution;
- `jit.off()` but without debugging hook. Speed of this mode is slowed N times that basis, but still is acceptable;
- `jit.off() with `debug.sethook`. This is the mode we use in `luadebug.lua` and which is showing the maximum performance impact which we could introduce in computing sensitive algorithms. 

Low-hanging fruit - memoization
------------------------------------

Initial versions of a `perf` profiles we have seen during this evaluation showed strange, excessive activity calling `string.gsub`. The problem was that each call to debugging hook reevaluated `normalize_path()` value for the file name, which left unchanged between calls (the name of a script itself). There is no point to calculate if thousands times - we could simply memoize calculated value in the hash:

https://gist.github.com/tsafin/da66ac0c992c6c4442d6e5572ae6882a#file-mandelbrot-dbg-3-lua-L44-L50
```lua
local normalized_paths = {}

local function normalize_path(file)
    local decorated_name = file
    if normalized_paths[decorated_name] ~= nil then
        return normalized_paths[decorated_name]
    end
    ...
```

Having this simple optimization, we could receive this performance improvement for hook enabled case - from this:
```
15:39 $ ./build/src/tarantool mandelbrot-dbg-2.lua 300
P4
300 300
jit.on(), no hook:      0.008785
jit.on(), line hook:    0.010473
jit.off(), no hook:     0.037613
jit.off(), line hook:   52.134762
```
To this:
```
15:40 $ ./build/src/tarantool mandelbrot-dbg-2.lua 300
P4
300 300
jit.on(), no hook:      0.00771
jit.on(), line hook:    0.007294
jit.off(), no hook:     0.040278
jit.off(), line hook:   8.310603
```

Different order of checks
---------------------------

As Gleb has asked, we have compared 2 implementations of `has_breakpoint*()` function - one which is checking (line, file) and one which is checking (file, line). 

The faster "common sense" version:
https://gist.github.com/tsafin/da66ac0c992c6c4442d6e5572ae6882a#file-mandelbrot-dbg-3-lua-L72-L74
```lua
local function has_breakpoint(file, line)
    return breakpoints[line] and breakpoints[line][file]
end
```

And slower, straightforward version:
https://gist.github.com/tsafin/da66ac0c992c6c4442d6e5572ae6882a#file-mandelbrot-dbg-3-lua-L76-L78
```lua
local function has_breakpoint_slow(file, line)
    return breakpoints[file] and breakpoints[file][line]
end
```

In the majority of cases, "common sense" version provided slightly better results. Please compare "lit.off(), line hooks" and "jit.off(), line hooks slow" below...

Collected data
----------------

After multiple reruns, and with algorithm fine-tuned, we have got to this data collected in a series of runs. 

<details>
  <summary>Log (expand me)</summary>
  
```
23:43 $ src/tarantool ../mandelbrot-dbg-3.lua 50
P4
50 50
jit.on(), no hook:      0.000349
jit.on(), line hook:    0.248944
jit.off(), no hook:     0.001033
jit.off(), line hook:   0.274717
jit.off(), line hook slow:      0.270286

23:48 $ src/tarantool ../mandelbrot-dbg-3.lua 100
P4
100 100
jit.on(), no hook:      0.001105
jit.on(), line hook:    0.926218
jit.off(), no hook:     0.003953
jit.off(), line hook:   2.073833
jit.off(), line hook slow:      2.81282

23:48 $ src/tarantool ../mandelbrot-dbg-3.lua 100
P4
100 100
jit.on(), no hook:      0.000973
jit.on(), line hook:    1.034362
jit.off(), no hook:     0.00566
jit.off(), line hook:   1.879252
jit.off(), line hook slow:      1.784875

00:07 $ src/tarantool ../mandelbrot-dbg-3.lua 200
P4
200 200
jit.on(), no hook:      0.005046
jit.on(), line hook:    5.472378
jit.off(), no hook:     0.025735
jit.off(), line hook:   6.25484
jit.off(), line hook slow:      6.548072

00:08 $ src/tarantool ../mandelbrot-dbg-3.lua 300
P4
300 300
jit.on(), no hook:      0.00722
jit.on(), line hook:    11.465756
jit.off(), no hook:     0.073952
jit.off(), line hook:   14.081694
jit.off(), line hook slow:      14.062397

00:09 $ src/tarantool ../mandelbrot-dbg-3.lua 400
P4
400 400
jit.on(), no hook:      0.014397
jit.on(), line hook:    19.553198
jit.off(), no hook:     0.092380000000002
jit.off(), line hook:   22.749138
jit.off(), line hook slow:      22.0638

00:10 $ src/tarantool ../mandelbrot-dbg-3.lua 500
P4
500 500
jit.on(), no hook:      0.020966
jit.on(), line hook:    27.183005
jit.off(), no hook:     0.131402
jit.off(), line hook:   34.398509
jit.off(), line hook slow:      37.348107

00:12 $ src/tarantool ../mandelbrot-dbg-3.lua 500
P4
500 500
jit.on(), no hook:      0.021133
jit.on(), line hook:    28.592071
jit.off(), no hook:     0.132882
jit.off(), line hook:   32.585221
jit.off(), line hook slow:      34.701433

00:14 $ src/tarantool ../mandelbrot-dbg-3.lua 750
P4
750 750
jit.on(), no hook:      0.046893
jit.on(), line hook:    63.915959
jit.off(), no hook:     0.30994500000001
jit.off(), line hook:   78.326954
jit.off(), line hook slow:      83.330058

00:18 $ src/tarantool ../mandelbrot-dbg-3.lua 1000
P4
1000 1000
jit.on(), no hook:      0.086248
jit.on(), line hook:    123.210141
jit.off(), no hook:     0.560276
jit.off(), line hook:   146.251999
jit.off(), line hook slow:      185.103624
```
</details>

![image](https://user-images.githubusercontent.com/4956130/211944550-1c82e41c-fc3c-4cef-b24e-86a7274d59f5.png)

![image](https://user-images.githubusercontent.com/4956130/211944572-40868f90-a8b2-4d15-8fc1-2748687a8f87.png)

Conclusions/observations
=================

The theory that debugger hook activated could lead to 100x time slowdown was both correct and wrong. We were correct assuming that slowdown would be significant, but we were too optimistic here, and actual ration is thousands times (i.e. 1000x - 2000x) and not hundreds.

We could try to reduce Lua overhead in debug hook via some tricks, and take care of a hot path efficiency, but (as `perf` profiles show) there is architectural limitation of `debug.getinfo` which has to create a lot of dynamic data about source location (i.e. `source`, `short_src`, `currentline`) which leads to significant GC pressure, and may take the biggest part of a CPU profile. 

> This problem might be worked-around via special LuaC code returning same debugging information directly, not in 
> generic table, but we rather not take this vendor specific way (as it would provide us in optimistic scenario 10x gain at 
> most), but instead would port in our implementation HALT patch introducing asynchronous `debug.sethalt` callback, 
> which would remove all line hook overhead altogether.

As we have proved, "common sense" (line, file) implementation of `has_breakpoint` is faster than straightforward (file, line). So on the hot path inside of debugger hook, we will continue to use "common sense" order. 

But for `bl` command, to show list of breakpoints in their natural order, we will continue to save the same data in `(file, line)` order, as it's not taking much space and cheap.

Reply:
After this benchmark, as I've discovered few inefficiencies - I've realized I need to tweak slightly code of a line hook, to make it run 10x times faster...

```diff
diff --git a/third_party/lua/luadebug.lua b/third_party/lua/luadebug.lua
index d339571e8..7159e28ab 100644
--- a/third_party/lua/luadebug.lua
+++ b/third_party/lua/luadebug.lua
@@ -157,17 +157,26 @@ local function format_stack_frame_info(info)
     return format_loc(source, info.currentline) .. " in " .. namewhat .. " " .. name
 end
 
+local normalized_paths = {}
+
+-- This function is on a hot path of a debugger hook.
+-- Try very hard to avoid wasting of CPU cycles, so reuse
+-- previously calculated results via memoization
 local function normalize_path(file)
+    local decorated_name = file
+    if normalized_paths[decorated_name] ~= nil then
+        return normalized_paths[decorated_name]
+    end
     -- If the name doesn't start with `@`, assume it's a file name if it's all
     -- on one line.
     if string.find(file, "^@") or not string.find(file, "[\r\n]") then
-        file = string.gsub(string.gsub(file, "^@", ""), "\\", "/")
-
-        file = fio.basename(file)
+        file = fio.basename(string.gsub(file, "^@", ""))
 
         -- some file systems allow newlines in file names; remove these.
         file = string.gsub(file, "\n", ' ')
     end
+    -- memoize current result - it will be needed very soon
+    normalized_paths[decorated_name] = file
     return file
 end
 
@@ -226,6 +235,12 @@ local function hook_factory(level)
     end
 end
 
+-- For the breakpoint hook we will bailout of a function
+-- for a majority of a calls (say in 99.999% of a cases).
+-- So it's very important to spend as small as possible time
+-- here: quick return if irrelevant, allocate the least possible
+-- memory, memoize calculated values - as for neighborhood lines it will
+-- be mostly the same
 local function hook_check_bps()
     return function(event, _)
         -- Skip events that don't have line information.
```

Reply:
Thank you for the benchmark! While it doesn't cover the case that I was worried about, it might not matter as 99% of all lines in all files are not breakpointed and the optimization is about this majority. The results seem inconclusive, as for the tests <= 500 they lie in the margin of error. As for the larger test, they seem to have just a bigger margin of error, because I got the following results when tested on a non-loaded cloud VM:
```
P4
500 500
jit.on(), no hook: 	0.029354
jit.on(), line hook: 	0.027187
jit.off(), no hook: 	0.144481
jit.off(), line hook: 	76.967195
jit.off(), line hook slow: 	77.480279
```
```
P4
650 650
jit.on(), no hook: 	0.048617
jit.on(), line hook: 	0.047987
jit.off(), no hook: 	0.241891
jit.off(), line hook: 	127.688481
jit.off(), line hook slow: 	127.559171
```
```
P4
750 750
jit.on(), no hook: 	0.06556
jit.on(), line hook: 	0.061379
jit.off(), no hook: 	0.329453
jit.off(), line hook: 	171.046775
jit.off(), line hook slow: 	172.998997
```
```
P4
1000 1000
jit.on(), no hook: 	0.118352
jit.on(), line hook: 	0.112927
jit.off(), no hook: 	0.57615
jit.off(), line hook: 	313.575491
jit.off(), line hook slow: 	309.270194
```